src/nvim/undo.c

```diff
@@ -1615,20 +1615,26 @@ void u_read_undo(char *name, const uint8_t *hash, const char *orig_name FUNC_ATT
   // Now that we have read the undo info successfully, free the current undo
   // info and use the info from the file.
   u_blockfree(curbuf);
-  curbuf->b_u_oldhead = old_idx < 0 ? NULL : uhp_table[old_idx];
-  curbuf->b_u_newhead = new_idx < 0 ? NULL : uhp_table[new_idx];
-  curbuf->b_u_curhead = cur_idx < 0 ? NULL : uhp_table[cur_idx];
-  curbuf->b_u_line_ptr = line_ptr;
-  curbuf->b_u_line_lnum = line_lnum;
-  curbuf->b_u_line_colnr = line_colnr;
-  curbuf->b_u_numhead = num_head;
-  curbuf->b_u_seq_last = seq_last;
-  curbuf->b_u_seq_cur = seq_cur;
-  curbuf->b_u_time_cur = seq_time;
-  curbuf->b_u_save_nr_last = last_save_nr;
-  curbuf->b_u_save_nr_cur = last_save_nr;
-
-  curbuf->b_u_synced = true;
+  u_state_T undostate = {
+    .buf = curbuf,
+    .save_b_u_synced = true,
+    .save_b_u_time_cur = seq_time,
+    .save_b_u_seq_cur = seq_cur,
+    .save_b_u_seq_last = seq_last,
+    .save_b_u_newhead = new_idx < 0 ? NULL : uhp_table[new_idx],
+    .save_b_u_oldhead = old_idx < 0 ? NULL : uhp_table[old_idx],
+    .save_b_u_curhead = cur_idx < 0 ? NULL : uhp_table[cur_idx],
+    .save_b_u_numhead = num_head,
+    .save_b_u_line_ptr = line_ptr,
+    .save_b_u_line_lnum = line_lnum,
+    .save_b_u_line_colnr = line_colnr,
+    .save_b_p_ul = curbuf->b_p_ul,
+    .save_b_changed = curbuf->b_changed,
+    .save_changedtick = buf_get_changedtick(curbuf),
+    .save_b_u_save_nr_cur = last_save_nr,
+    .save_b_u_save_nr_last = last_save_nr,
+  };
+  u_restore_state(&undostate);
```

Comment:
What is the point of this change? It seems to make code more confusing.

Reply:
The goal was to add a generic interface to save/restore the undo state.

Reply:
The `save_*` prefix on each field seems redundant?

Reply:
> The `save_*` prefix on each field seems redundant?

Agree. And if we remove the prefix, we could use it to replace the `b_u_*`` loose fields from `file_buffer`.

https://github.com/neovim/neovim/blob/ccf328172bac2b02f9bd19fa58e105958514a28a/src/nvim/buffer_defs.h#L568-L583

Reply:
> > The `save_*` prefix on each field seems redundant?
> 
> Agree. And if we remove the prefix, we could use it to replace the `b_u_*` loose fields from `file_buffer`.
> 
> https://github.com/neovim/neovim/blob/ccf328172bac2b02f9bd19fa58e105958514a28a/src/nvim/buffer_defs.h#L568-L583

I like that idea a lot. Do you plan on making a PR for it?

Reply:
I also thought these `b_u_` fields can moved into a structs with slightly different names, and then `#define`s can be added for the original names, but I had some trouble giving a name to the struct and its fields.